### DISCUSSION TITLE
Generate GDJS types

### DISCUSSION THREAD

**Discussion by arthuro555** [4ian/GDevelop#2802] [Enhancement ideas (and their technical discussions)]
Instead of bundling both the built and non-built runtime for autocompletion, we could generate type definitions with tsc and bundle those with GDevelop for autocompletions. This should help make GDevelop more lightweight (and I must admit that my primary motivation is that this would also be pretty useful for type checking external tools that use the gdjs api, like my modding tool gdmod and other tool ideas I have)

**Comment by 4ian** [comment]
Yes! That's even a comment in the codebase ;) 
https://github.com/4ian/GDevelop/blob/f3a49ad2cfd48efa6bed17565041094390a199b6/newIDE/app/src/CodeEditor/LocalCodeEditorAutocompletions.js#L40-L47

By doing this, we can remove the copy of the sources (which is now done anyway only when starting or bundling the IDE). But we have to change the TS config to export these type definitions (takes a bit more time for tsc) and output them in a place where read by the IDE (and ensure this is done before bundling).

**Reply by arthuro555** [reply]
If it increases tsc "build" time that much, maybe we can have two tsconfigs and two npm scripts running tsc with those two configs (`check-types` and `generate-dts`), with one exclusively for type checking and the other for also emitting declarations.
